### PR TITLE
docs: Describe arbitrary metadata logging

### PR DESCRIPTION
**Pull request by tara-det-ai**
## Docs Updated

- [x] Initial draft reviewed by ET team
- [x] [Tutorial added](https://determined-ai-docs.s3.us-west-2.amazonaws.com/previews/c7d677a029ed29020a6c706c715ecc6c/tutorials/metadata-logging.html)
- [x] [Exp Config Ref updated](https://determined-ai-docs.s3.us-west-2.amazonaws.com/previews/c7d677a029ed29020a6c706c715ecc6c/reference/experiment-config-reference.html#metadata-logging)
- [x] [WebUI page ](https://determined-ai-docs.s3.us-west-2.amazonaws.com/previews/c7d677a029ed29020a6c706c715ecc6c/tools/webui-if.html#viewing-and-filtering-metadata)updated

## Description

This feature provides users with the ability to log comprehensive metadata for each experiment run, enabling them to capture and record detailed context and parameters specific to their project needs. In addition, it empowers users to filter and analyze experiment runs based on custom metadata key-value pairs, facilitating more targeted and efficient evaluation of model performance and outcomes.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] New features have been approved by the corresponding PM
- [ ] User-facing API changes have the "User-facing API Change" label
- [ ] Release notes have been added as a separate file under `docs/release-notes/`
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses have been included for new code which was copied and/or modified from any external code

[DET-123]: https://hpe-aiatscale.atlassian.net/browse/DET-123?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: docs/tools/webui-if.rst [hunk 1]
```diff
@@ -241,3 +241,26 @@ Clear the message with the following command:
...
+********************************
+ Viewing and Filtering Metadata
+********************************
+
+You can use the WebUI to view and filter experiments based on logged metadata. For a tutorial on how
+to log metadata, visit :ref:`metadata-logging-tutorial`.
+
+-  In the Overview tab of the experiment, you can filter and sort runs based on metadata values
+   using the filter menu.
+-  In the Runs (Table) view, metadata columns are displayed alongside other experiment information.
```
Comment (AmanuelAaron) [inline docs/tools/webui-if.rst:254]: "experiment" -> "run"

File: docs/tutorials/metadata-logging.rst [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+2. Click on the trial you want to inspect.
+3. In the trial details page, find the "Metadata" section under the "Overview" tab.
+
+***********************************
+ Filtering and Sorting by Metadata
+***********************************
+
+The :ref:`Web UI <web-ui-if>` allows you to filter and sort experiments based on logged metadata:
+
+1. Navigate to the Experiments List page in the WebUI.
```
Comment (AmanuelAaron) [inline docs/tutorials/metadata-logging.rst:73]: "Experiments List" -> "Runs Table"

File: docs/tutorials/metadata-logging.rst [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+1. Open the WebUI and navigate to your experiment.
+2. Click on the trial you want to inspect.
+3. In the trial details page, find the "Metadata" section under the "Overview" tab.
+
+***********************************
+ Filtering and Sorting by Metadata
+***********************************
+
+The :ref:`Web UI <web-ui-if>` allows you to filter and sort experiments based on logged metadata:
```
Comment (AmanuelAaron) [inline docs/tutorials/metadata-logging.rst:71]: "experiments" -> "runs"

File: docs/tutorials/metadata-logging.rst [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+3. Analyze the impact of software versions on performance.
+
+Summary
+=======
+
+Arbitrary Metadata Logging enhances your experiment tracking capabilities by allowing you to:
+
+1. Log custom metadata specific to your experiments.
+2. View logged metadata in the WebUI for each trial.
+3. Filter and sort experiments based on custom metadata.
```
Comment (AmanuelAaron) [inline docs/tutorials/metadata-logging.rst:125]: experiments -> runs

**Review by AmanuelAaron — COMMENTED** [review]
Metadata is logged, filtered & sorted on the run level

File: docs/reference/experiment-config-reference.rst [hunk 5]
```diff
@@ -1654,3 +1654,13 @@ If :ref:`gres_supported <cluster-configuration-slurm>` is set to ``false``, spec
...
+
+******************
+ Metadata Logging
+******************
+
+Determined supports logging arbitrary metadata for experiments. This feature allows users to store
+additional context and information about their runs. To log metadata, use the following function in
+your code:
+
+TO DO: ADD EXAMPLE USAGE
```
Comment (tara-det-ai) [inline docs/reference/experiment-config-reference.rst:1666]: @AmanuelAaron do we need any content for the experiment config ref page?

File: docs/tutorials/metadata-logging.rst [hunk 6]
```diff
@@ -0,0 +1,137 @@
+.. _metadata-logging-tutorial:
+
+############################
+ Arbitrary Metadata Logging
+############################
+
+This tutorial demonstrates how to use Arbitrary Metadata Logging in Determined AI to log custom metadata for your experiments.
```
Comment (tara-det-ai) [inline docs/tutorials/metadata-logging.rst:7]: remove "Determined AI"

File: docs/tutorials/metadata-logging.rst [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+When using Arbitrary Metadata Logging, consider the following:
+
+- Metadata is stored efficiently for fast retrieval and filtering.
+- Avoid logging very large metadata objects, as this may impact performance.
+- Use consistent naming conventions for keys to make filtering and sorting easier.
+- For deeply nested JSON structures, filtering and sorting are supported at the top level.
+
+Example Use Case
+================
```
Comment (tara-det-ai) [inline docs/tutorials/metadata-logging.rst:92]: @keita-determined @AmanuelAaron this is placeholder text and the code shown is just for placeholder purposes. Do we need an example use case? If so, what are the proper steps?

File: docs/tools/webui-if.rst [hunk 8]
```diff
@@ -241,3 +241,26 @@ Clear the message with the following command:
...
+-  On the Run details page, you'll find the "Metadata" section under the "Overview" tab, displaying
+   all logged metadata for that run.
+-  To download the metadata in JSON format, click the "Download" button.
+
+To filter runs based on metadata:
+
+#. In the Runs view, click on the filter icon.
+#. Select a metadata field from the dropdown menu.
+#. Choose a condition (is, is not, or contains) and enter a value.
+#. Click "Apply" to filter the runs based on the metadata.
```
Comment (keita-determined) [inline docs/tools/webui-if.rst:264]: i dont think there's apply button

File: docs/tutorials/metadata-logging.rst [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+ Filtering and Sorting by Metadata
+***********************************
+
+The :ref:`Web UI <web-ui-if>` allows you to filter and sort experiments based on logged metadata:
+
+1. Navigate to the Experiments List page in the WebUI.
+2. Click on the filter icon.
+3. Select a metadata field from the dropdown menu.
+4. Choose a condition (is, is not, or contains) and enter a value.
+5. Click "Apply" to filter the experiments based on the metadata.
```
Comment (keita-determined) [inline docs/tutorials/metadata-logging.rst:77]: same here

File: docs/tools/webui-if.rst (hunk 8 above)
Comment (tara-det-ai) [reply docs/tools/webui-if.rst:264]: you are right, thank you!

File: docs/tutorials/metadata-logging.rst (hunk 7 above)
Comment (tara-det-ai) [reply docs/tutorials/metadata-logging.rst:92]: i'll use the PostRunMetadata example from the demo

**Pull request merged**